two inputs possible

mpahl opened the pull request.

mpahl committed:
two inputs possible

hansendx commented:
Please be a little more elaborate with your pull request description and commit messages.

hansendx commented on collect_stata/write_json.py:
```diff
@@ -202,10 +202,44 @@ def generate_statistics(
     data: pd.DataFrame,
-    metadata: List[Dict[str, Union[str, Dict[str, List[Union[int, str, bool]]]]]],
+    metadata_en: List[Dict[str, Union[str, Dict[str, List[Union[int, str, bool]]]]]],
+    metadata_de: List[Dict[str, Union[str, Dict[str, List[Union[int, str, bool]]]]]],
```
Why do you pass this to write_json and not directly into update_meatadata?

hansendx commented on collect_stata/write_json.py:
```diff
@@ -202,10 +202,44 @@ def generate_statistics(
 
+def update_metadata(
+    metadata: List[Dict[str, Union[str, Dict[str, List[Union[int, str, bool]]]]]],
+    metadata_de: List[Dict[str, Union[str, Dict[str, List[Union[int, str, bool]]]]]],
```
We updated to python 3.8 on the develop branch. You can use TypedDicts to make this more readable and more precise. https://mypy.readthedocs.io/en/latest/more_types.html#typeddict

hansendx commented on collect_stata/write_json.py:
```diff
@@ -202,10 +202,44 @@ def generate_statistics(
+    metadata_de: Metadata of the german imported data.
+
+    Output:
+    metadata: Metadata variable with german and english labels if given.
```
Please use Args: and Returns. Please indent blocks after `:`
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

hansendx commented on collect_stata/write_json.py:
```diff
@@ -202,10 +202,44 @@ def generate_statistics(
 
     return metadata
 
+def update_metadata(
```
This would better fit into some of the other modules.
https://en.wikipedia.org/wiki/Cohesion_(computer_science)

mpahl committed:
fix typing errors

mpahl committed:
update _run function

hansendx commented:
Please integrate the changes from develop.
The function stata_to_json was replaced with the class StataToJson.

mpahl committed:
Merge branch 'develop' into rewrite_for_bilingual_datasets

hansendx commented on collect_stata/__main__.py:
```diff
@@ -4,7 +4,7 @@
 import logging
 import os
-import pathlib
+import pathlib import Path
```
syntax error

hansendx committed:
Fix syntax error

hansendx committed:
Fix KeyError

Sometimes elem["categories"]["missings"] was not
initiated before calling
elem["categories"]["missings"].append()

hansendx committed:
Fix TypeError when writing json

variable_meta["categories"]["values"] sometimes contained
numbers of numpy integer types.
Those numbers caused json.dumps() to fail since
it could not be serialize these integer objects.

hansendx committed:
Fix wrong indentation

Wrong indentation in multithreading block caused
only one process to run.

hansendx committed:
Update dependencies

hansendx committed:
Fix multiprocessing with only one language

Process gathering and starting was only done
inside the block for processing two languages.

* Remove lint
* Fix type hints
* Refactor bloated code

hansendx committed:
Simplify identification of missing values

hansendx committed:
Refactor __main__ and write_json

* Homogenize naming
* Fix formatting

hansendx committed:
Fix value label reading for multi language files

* Value labels are not switched when language of
  the stata file is switched.

  + Variable labels are though, which lead to the
    false expectations and faulty implementation.

* Value labels are bound to a key which in turn is
  associated with a language and a variable.
  This key seems to be arbitrarily assignable.

* It seems that value labels can only be associated
  with their variables through the position of the
  key inside the lbllist stata reader attribute.

hansendx committed:
Fix datatype issues

* Cast statistics values from numpy internal type
  to python numeric values.

hansendx commented:
labels_de list is too short in some variables.
It seems that all other lists under the "categories" key are filled in to make them comparable with the value label lists of related variables. 

Here is a snippet with labels and labels_de:

```json
      "labels": [
        "[-6] Version of questionnaire with modified filtering",
        "[-5] Not included in this version of the questionnaire",
        "[-4] Inadmissible multiple response",
        "[-3] Answer improbable",
        "[-2] Does not apply",
        "[-1] No Answer",
        "[0] 0 Satisfied: On Scale 0-Low to 10-High",
        "[10] 10 Satisfied: On Scale 0-Low to 10-High",
        "8",
        "7",
        "5",
        "9",
        "6",
        "4",
        "3",
        "2",
        "1"
      ],
      "labels_de": [
        "[-6] Fragebogenversion mit geaenderter Filterfuehrung",
        "[-5] In Fragebogenversion nicht enthalten",
        "[-4] Unzulaessige Mehrfachantwort",
        "[-3] nicht valide",
        "[-2] trifft nicht zu",
        "[-1] keine Angabe",
        "[0] Ganz unzufrieden",
        "[10] Ganz zufrieden"
      ],
```

Here is the full variable:
```json
  {
    "name": "qp14301",
    "dataset": "qp",
    "label": "Satisfaction With Life At Today",
    "categories": {
      "values": [
        -6,
        -5,
        -4,
        -3,
        -2,
        -1,
        0,
        10,
        8,
        7,
        5,
        9,
        6,
        4,
        3,
        2,
        1
      ],
      "labels": [
        "[-6] Version of questionnaire with modified filtering",
        "[-5] Not included in this version of the questionnaire",
        "[-4] Inadmissible multiple response",
        "[-3] Answer improbable",
        "[-2] Does not apply",
        "[-1] No Answer",
        "[0] 0 Satisfied: On Scale 0-Low to 10-High",
        "[10] 10 Satisfied: On Scale 0-Low to 10-High",
        "8",
        "7",
        "5",
        "9",
        "6",
        "4",
        "3",
        "2",
        "1"
      ],
      "labels_de": [
        "[-6] Fragebogenversion mit geaenderter Filterfuehrung",
        "[-5] In Fragebogenversion nicht enthalten",
        "[-4] Unzulaessige Mehrfachantwort",
        "[-3] nicht valide",
        "[-2] trifft nicht zu",
        "[-1] keine Angabe",
        "[0] Ganz unzufrieden",
        "[10] Ganz zufrieden"
      ],
      "missings": [
        true,
        true,
        true,
        true,
        true,
        true,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false
      ],
      "frequencies": [
        0,
        0,
        0,
        0,
        0,
        66,
        111,
        1554,
        7484,
        5242,
        2993,
        2828,
        2674,
        753,
        512,
        269,
        90
      ]
    },
    "scale": "cat",
    "label_de": "Lebenszufriedenh. gegenwaertig",
    "study": "soep-core",
    "statistics": {
      "valid": 24510,
      "invalid": 66
    }
  },
  {
    "name": "qp14302",
    "dataset": "qp",
    "label": "Satisfaction With Life In Five Years",
    "categories": {
      "values": [
        -6,
        -5,
        -4,
        -3,
        -2,
        -1,
        0,
        10,
        8,
        7,
        9,
        5,
        6,
        4,
        3,
        2,
        1
      ],
      "labels": [
        "[-6] Version of questionnaire with modified filtering",
        "[-5] Not included in this version of the questionnaire",
        "[-4] Inadmissible multiple response",
        "[-3] Answer improbable",
        "[-2] Does not apply",
        "[-1] No Answer",
        "[0] 0 Satisfied: On Scale 0-Low to 10-High",
        "[10] 10 Satisfied: On Scale 0-Low to 10-High",
        "8",
        "7",
        "9",
        "5",
        "6",
        "4",
        "3",
        "2",
        "1"
      ],
      "labels_de": [
        "[-6] Fragebogenversion mit geaenderter Filterfuehrung",
        "[-5] In Fragebogenversion nicht enthalten",
        "[-4] Unzulaessige Mehrfachantwort",
        "[-3] nicht valide",
        "[-2] trifft nicht zu",
        "[-1] keine Angabe",
        "[0] Ganz unzufrieden",
        "[10] Ganz zufrieden"
      ],
      "missings": [
        true,
        true,
        true,
        true,
        true,
        true,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false,
        false
      ],
      "frequencies": [
        0,
        0,
        0,
        0,
        0,
        560,
        141,
        1836,
        6847,
        4347,
        3732,
        2799,
        2359,
        924,
        580,
        328,
        123
      ]
    },
    "scale": "cat",
    "label_de": "Lebenszufriedenh. in 5 Jahren",
    "study": "soep-core",
    "statistics": {
      "valid": 24016,
      "invalid": 560
    }
  }
```

hansendx committed:
Update with orphaned changes

hansendx committed:
Fix attribute name

hansendx committed:
Update for newer pandas version

hansendx committed:
Fix issue where german labels are dropped
